core/src/main/scala/kafka/log/Log.scala

```diff
@@ -85,6 +85,8 @@ object LogAppendInfo {
  * @param validBytes The number of valid bytes
  * @param offsetsMonotonic Are the offsets in this message set monotonically increasing
  * @param lastOffsetOfFirstBatch The last offset of the first batch
+ * @param leaderHWIncremented true if the high watermark is increased when appending record. Otherwise, false.
+ *                            this field is updated after appending record so the default value is None.
```

Comment:
Perhaps we can make this a bit clearer. Sth like the following.

leaderHWIncremented has 3 possible values: (1) If records are not appended due to an error, the value will be None; (2) if records are appended successfully and HWM is advanced, the value is Some(true); (3) if records are appended successfully and HWM is not advanced, the value is Some(false).

Reply:
@chia7712 I suggest using a sealed trait with 3 case objects to make this clearer. Using `Option[Boolean]` as a tristate value is generally best avoided.

Reply:
Will copy that